<center style = "font-family: Trebuchet MS, sans-serif;">
    <h1 style = "background-color:#D8BFD8; color:white;"> MACHINE LEARNING - LAB 04</h1>
    <hr style = "background-color:#D8BFD8;">
</center>
<h3 style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;"> 
    <b>NAME: </b>Sondos Suhail <br>
    <b>ID: </b>20K-1057
</h3>
<hr style = "background-color:#D8BFD8;">

<h3 style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;"> Imported Libraries </h3>

In [1]:
import pandas as pd
import numpy as np
from scipy.io import arff
import statistics
import math
from sklearn.model_selection import cross_val_score, train_test_split, StratifiedShuffleSplit, RepeatedKFold
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier

<h3  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;"> TASK 1 </h3>

<hr >
<h4  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;"> BREAST CANCER DATASET DESCRIPTION </h4>

<ol  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;">
    <li>Sample Code Number : ID No. </li>
   <li>Clump Thickness: 1 - 10 </li>
   <li>Uniformity of Cell Size: 1 - 10 </li>
   <li>Uniformity of Cell Shape: 1 - 10 </li>
   <li>Marginal Adhesion: 1 - 10 </li>
   <li>Single Epithelial Cell Size: 1 - 10 </li>
   <li>Bare Nuclei: 1 - 10 </li>
   <li>Bland Chromatin: 1 - 10 </li>
   <li>Normal Nucleoli: 1 - 10 </li>
  <li>Mitoses: 1 - 10 </li>
  <li>Class: <ol type = "i"> <li> 0 - Benign </li> <li> 1 - Malignant </li> </ol> </li>
</ol>

In [20]:
names = ["Sample ID", "Clump Thickness", "Uniformity of Cell Size", "Uniformity of Cell Shape",
         "Marginal Adhesion", "Single Epithelial Cell Size", "Bare Nuclei", "Bland Chromatin", 
         "Normal Nucleoli", "Mitoses", "Class"]   

to_replace = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10"]
df1 = pd.read_csv("breast-cancer.data", names = names)
df1 = df1.drop("Sample ID", axis = 1)
df1 = df1.replace('?', 0)
for i in range(1, 11):
    df1["Bare Nuclei"] = df1["Bare Nuclei"].replace(to_replace[i-1], i)

df1["Class"] = df1["Class"].replace(2, 0)
df1["Class"] = df1["Class"].replace(4, 1)
df1

Clump Thickness  Uniformity of Cell Size  Uniformity of Cell Shape  \
0                  5                        1                         1   
1                  5                        4                         4   
2                  3                        1                         1   
3                  6                        8                         8   
4                  4                        1                         1   
..               ...                      ...                       ...   
694                3                        1                         1   
695                2                        1                         1   
696                5                       10                        10   
697                4                        8                         6   
698                4                        8                         8   

     Marginal Adhesion  Single Epithelial Cell Size  Bare Nuclei  \
0                    1                            2            1   
1                    5                            7           10   
2                    1                            2            2   
3                    1                            3            4   
4                    3                            2            1   
..                 ...                          ...          ...   
694                  1                            3            2   
695                  1                            2            1   
696                  3                            7            3   
697                  4                            3            4   
698                  5                            4            5   

     Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0                  3                1        1      0  
1                  3                2        1      0  
2                  3                1        1      0  
3                  3                7        1      0  
4                  3                1        1      0  
..               ...              ...      ...    ...  
694                1                1        1      0  
695                1                1        1      0  
696                8               10        2      1  
697               10                6        1      1  
698               10                4        1      1  

[699 rows x 10 columns]

<h3  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;">Holdout - Repeated 100 Times</h3>

In [21]:
array = df1.values
X = array[:, 0:9]
Y = array[:, 9]

CV2 = StratifiedShuffleSplit(n_splits = 100, test_size = 0.3, random_state = 0)

In [22]:
dt = DecisionTreeClassifier( criterion = 'gini')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Gini - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Gini - Without Pruning
Accuracy:  93.91 %
Standard Deviation:  1.36


In [23]:
dt = DecisionTreeClassifier( criterion = 'gini', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Gini - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))


Gini - With Pruning
Accuracy:  93.8 %
Standard Deviation:  1.76


In [24]:
dt = DecisionTreeClassifier( criterion = 'entropy')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Entropy - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))


Entropy - Without Pruning
Accuracy:  93.79 %
Standard Deviation:  1.39


In [25]:
dt = DecisionTreeClassifier( criterion = 'entropy', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Entropy - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))


Entropy - With Pruning
Accuracy:  94.3 %
Standard Deviation:  1.46


<h3  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;"> 10 x 10 Fold </h3>

In [26]:
array = df1.values
X = array[:, 0:9]
Y = array[:, 9]

CV = RepeatedKFold(n_splits = 10, n_repeats = 10, random_state = 1)

In [27]:
dt = DecisionTreeClassifier( criterion = 'gini')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Gini - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Gini - Without Pruning
Accuracy:  93.76 %
Standard Deviation:  2.45


In [29]:
dt = DecisionTreeClassifier( criterion = 'gini', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Gini - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Gini - With Pruning
Accuracy:  93.43 %
Standard Deviation:  2.53


In [30]:
dt = DecisionTreeClassifier( criterion = 'entropy')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Entropy - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))


Entropy - Without Pruning
Accuracy:  93.62 %
Standard Deviation:  2.87


In [32]:
dt = DecisionTreeClassifier( criterion = 'entropy', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Entropy - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Entropy - With Pruning
Accuracy:  94.05 %
Standard Deviation:  2.91


<hr >
<h4  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;">LEAF DATASET DESCRIPTION </h4>

<ol  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;">
    <li>Class (Species)</li>
    <li> Specimen Number</li>
    <li> Eccentricity</li>
    <li>Aspect Ratio</li>
    <li>Elongation</li>
    <li>Solidity</li>
    <li>Stochastic Convexity</li>
    <li>Isoperimetric Factor</li>
    <li>Maximal Indentation Depth </li>
    <li>Lobedness</li>
    <li>Average Intensity </li>
    <li>Average Contrast</li>
    <li>Smoothness</li>
    <li>Third moment</li>
    <li>Uniformity</li>
    <li>Entropy</li>
</ol>
<hr>
<p style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;">
The present database comprises 40 different plant species. Table 1 details each plant's scientifc
name and the number of leaf specimens available by species. Species numbered from 1 to 15
and from 22 to 36 exhibit simple leaves and species numbered from 16 to 21 and from 37 to 40
have complex leaves.
</p>
<img src = "Leaf 1.png" style = "width:500px; height:550px;">
<img src = "Leaf 2.png" style = "width:500px;height:550px;">
<img src = "Leaf 3.png" style = "width:450px;height:400px;">

In [160]:
names = [ "Class", "Specimen Number", "Eccentricity", "Aspect Ratio",
    "Elongation", "Solidity", "Stochastic Convexity", "Isoperimetric Factor", "Maximal Indentation Depth",
    "Lobedness", "Average Intensity", "Average Contrast", "Smoothness",
    "Third moment", "Uniformity", "Entropy"]

df2 = pd.read_csv("Leaf.csv", names = names)
df2 = df2.drop("Specimen Number", axis = 1)
df2 = df2.sample(frac = 1)
df2

Class  Eccentricity  Aspect Ratio  Elongation  Solidity  \
187     23       0.56840        1.1876     0.48001   0.91247   
79       8       0.98600        6.1763     0.83942   0.97837   
268     30       0.48030        1.2022     0.26797   0.94544   
244     28       0.91996        2.6668     0.63083   0.97894   
71       8       0.98487        6.0229     0.83628   0.96522   
..     ...           ...           ...         ...       ...   
259     29       0.78965        1.8186     0.45353   0.94929   
277     31       0.99767       12.0090     0.94529   0.67005   
155     14       0.93802        2.8087     0.65774   0.97091   
156     14       0.94058        3.0366     0.69248   0.93952   
201     24       0.67925        1.4870     0.33465   0.97307   

     Stochastic Convexity  Isoperimetric Factor  Maximal Indentation Depth  \
187               0.79474               0.38850                   0.089162   
79                0.99298               0.31607                   0.011769   
268               0.99825               0.52139                   0.022778   
244               1.00000               0.58002                   0.010311   
71                1.00000               0.29509                   0.024667   
..                    ...                   ...                        ...   
259               1.00000               0.63343                   0.030371   
277               0.94386               0.11114                   0.198980   
155               1.00000               0.58194                   0.006732   
156               0.99474               0.42979                   0.031219   
201               0.98246               0.77130                   0.012399   

     Lobedness  Average Intensity  Average Contrast  Smoothness  Third moment  \
187   1.446900           0.027113          0.101820    0.010261      0.003789   
79    0.025208           0.051048          0.135240    0.017961      0.006018   
268   0.094430           0.066259          0.148180    0.021485      0.006194   
244   0.019349           0.112500          0.197850    0.037671      0.011205   
71    0.110740           0.030300          0.097095    0.009339      0.002783   
..         ...                ...               ...         ...           ...   
259   0.167870           0.008049          0.043450    0.001884      0.000454   
277   7.206200           0.010215          0.051104    0.002605      0.000668   
155   0.008249           0.052218          0.134910    0.017874      0.005737   
156   0.177380           0.038691          0.107600    0.011445      0.003216   
201   0.027980           0.007600          0.044132    0.001944      0.000491   

     Uniformity  Entropy  
187    0.000107  0.65567  
79     0.000313  1.16600  
268    0.000735  1.36690  
244    0.000789  2.22410  
71     0.000183  0.85025  
..          ...      ...  
259    0.000029  0.36533  
277    0.000045  0.39905  
155    0.000302  1.21610  
156    0.000248  1.08430  
201    0.000037  0.28857  

[340 rows x 15 columns]

<h3  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;">Holdout - Repeated 100 Times</h3>

In [161]:
array = df2.values
X = array[:, 1:15]
Y = array[:, 0]

CV2 = StratifiedShuffleSplit(n_splits = 100, test_size = 0.3, random_state = 0)

In [162]:
dt = DecisionTreeClassifier( criterion = 'gini')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Gini - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Gini - Without Pruning
Accuracy:  61.75 %
Standard Deviation:  4.12


In [163]:
dt = DecisionTreeClassifier( criterion = 'gini', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Gini - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Gini - With Pruning
Accuracy:  58.35 %
Standard Deviation:  3.85


In [164]:
dt = DecisionTreeClassifier( criterion = 'entropy')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Entropy - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Entropy - Without Pruning
Accuracy:  61.14 %
Standard Deviation:  4.51


In [165]:
dt = DecisionTreeClassifier( criterion = 'entropy', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV2, n_jobs = -1)
print("Entropy - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Entropy - With Pruning
Accuracy:  60.94 %
Standard Deviation:  4.36


<h3  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;"> 10 x 10 Fold </h3>

In [166]:
array = df2.values
X = array[:, 1:15]
Y = array[:, 0]

CV = RepeatedKFold(n_splits = 10, n_repeats = 10, random_state = 1)

In [167]:
dt = DecisionTreeClassifier( criterion = 'gini')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Gini - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Gini - Without Pruning
Accuracy:  63.35 %
Standard Deviation:  8.44


In [168]:
dt = DecisionTreeClassifier( criterion = 'gini', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Gini - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Gini - With Pruning
Accuracy:  55.24 %
Standard Deviation:  8.95


In [169]:
dt = DecisionTreeClassifier( criterion = 'entropy')
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Entropy - Without Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Entropy - Without Pruning
Accuracy:  64.24 %
Standard Deviation:  8.93


In [170]:
dt = DecisionTreeClassifier( criterion = 'entropy', ccp_alpha = 0.015)
Acc_n_Std = cross_val_score(dt, X, Y, scoring = 'accuracy', cv = CV, n_jobs = -1)
print("Entropy - With Pruning")
print('Accuracy: ', round(np.mean(Acc_n_Std)*100, 2), "%")
print( 'Standard Deviation: ', round(np.std(Acc_n_Std)*100, 2))

Entropy - With Pruning
Accuracy:  64.18 %
Standard Deviation:  8.25


<p style = "text-align: center; font-size: 20px"><b><u>Holdout - Repeated 100 Times</u></b></p> <br>
<table style = "border: 1px solid white; border-collapse: collapse; width: 420px; height: 150px">
  <tr>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center"></th>
      <th colspan="4" style = "border: 1px solid white; border-collapse: collapse; text-align: center;font-size:14px">Datasets</th>
  </tr>
  <tr>
    <th style = "border: 1px solid white; border-collapse: collapse; text-align: center"></th>
    <th colspan = "2" style = "border: 1px solid white; border-collapse: collapse; text-align: center">Breast Cancer</th>
    <th colspan = "2" style = "border: 1px solid white; border-collapse: collapse; text-align: center">Leaf</th>
  </tr>
  <tr>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center"></th>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Standard Deviation</th>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Average Accuracy</th>
       <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Standard Deviation</th>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Average Accuracy</th>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Gini <br> (without pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">1.34</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">94.0 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">4.12</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center"> 61.75 %</td>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Gini <br> (with pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">1.75</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">93.82 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center"> 3.85</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">58.35 %</td>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Entropy <br> (without pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">1.46</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">93.88 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">4.51</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">61.14 %</td>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Entropy <br> (with pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">1.45</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">94.31 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">4.36</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">60.94 %</td>
  </tr>
</table>

<hr>
<br>
<p style = "text-align: center; font-size: 20px"><b><u>10 x 10 Fold </u></b></p> <br>
<table style = "border: 1px solid white; border-collapse: collapse; width: 420px; height: 150px">
  <tr>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center"></th>
      <th colspan="4" style = "border: 1px solid white; border-collapse: collapse; text-align: center;font-size:14px">Datasets</th>
  </tr>
  <tr>
    <th style = "border: 1px solid white; border-collapse: collapse; text-align: center"></th>
    <th colspan = "2" style = "border: 1px solid white; border-collapse: collapse; text-align: center">Breast Cancer</th>
    <th colspan = "2" style = "border: 1px solid white; border-collapse: collapse; text-align: center">Leaf</th>
  </tr>
  <tr>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center"></th>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Standard Deviation</th>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Average Accuracy</th>
       <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Standard Deviation</th>
      <th style = "border: 1px solid white; border-collapse: collapse; text-align: center">Average Accuracy</th>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Gini <br> (without pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">2.45</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">93.76 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">8.44</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">63.35 %</td>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Gini <br> (with pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">2.53</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">93.43 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">8.95</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">55.24 %</td>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Entropy <br> (without pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">2.87</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">93.62 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">8.93</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">64.24 %</td>
  </tr>
  <tr>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">DT Using Entropy <br> (with pruning)</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">2.91</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">94.05 %</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">8.25</td>
      <td style = "border: 1px solid white; border-collapse: collapse; text-align: center">64.18 %</td>
  </tr>
</table>

<h3  style = "color:#D8BFD8; font-family:Trebuchet MS, sans-serif;"> TASK 2 </h3>

In [127]:
#entropy = -yes*log(yes) - no*log(no)
def entropy(yes, no):
    if yes != 0 and no != 0:
        e = -(yes*np.log2(yes)) - (no*np.log2(no))
    else:
        return 0
    return e

In [159]:
data = {"Outlook" : ["Sunny", "Sunny", "Overcast", "Rain", "Rain", "Rain", "Overcast",
                     "Sunny", "Sunny", "Rain", "Sunny", "Overcast", "Overcast", "Rain"],
       "Temperature": ["Hot", "Hot", "Hot", "Mild", "Cool", "Cool", "Cool", "Mild",
                      "Cool", "Mild", "Mild", "Mild", "Hot", "Mild"],
       "Humidity" : ["High", "High", "High", "High", "Normal", "Normal", "Normal",
                     "High", "Normal", "Normal", "Normal", "High", "Normal", "High"],
       "Windy" : ["False", "True", "False", "False", "False", "True", "True",
                 "False", "False", "False", "True", "True", "False", "True"],
       "Play?" : ["No", "No", "Yes", "Yes", "Yes", "No", "Yes", "No", "Yes", "Yes", "Yes", "Yes", "Yes", "No"]
       }

df3 = pd.DataFrame(data)
total_count = df3.shape[0]
data_no = df3["Play?"].value_counts()['No']
data_yes = df3["Play?"].value_counts()['Yes']
no_prob = data_no/total_count
yes_prob = data_yes/total_count

play_entropy = entropy(yes_prob, no_prob)
      
print("Play Entropy: ", round(play_entropy, 5))
df3

Play Entropy:  0.94029


Outlook Temperature Humidity  Windy Play?
0      Sunny         Hot     High  False    No
1      Sunny         Hot     High   True    No
2   Overcast         Hot     High  False   Yes
3       Rain        Mild     High  False   Yes
4       Rain        Cool   Normal  False   Yes
5       Rain        Cool   Normal   True    No
6   Overcast        Cool   Normal   True   Yes
7      Sunny        Mild     High  False    No
8      Sunny        Cool   Normal  False   Yes
9       Rain        Mild   Normal  False   Yes
10     Sunny        Mild   Normal   True   Yes
11  Overcast        Mild     High   True   Yes
12  Overcast         Hot   Normal  False   Yes
13      Rain        Mild     High   True    No

In [152]:
data_outlook = df3.filter(["Outlook", "Play?"])

sunny = 0
sunny_yes = 0
sunny_no = 0
overcast = 0
overcast_yes = 0
overcast_no = 0
rain = 0
rain_yes = 0
rain_no = 0

array = data_outlook.values
outlook = array[: , 0]
play = array[: , 1]

for i in range(0, total_count):
    if outlook[i] == "Sunny":
        sunny += 1
        if play[i] == "Yes":
            sunny_yes += 1
        else:
            sunny_no += 1
    elif outlook[i] == "Overcast":
        overcast += 1
        if play[i] == "Yes":
            overcast_yes += 1
        else:
            overcast_no += 1
    else:
        rain += 1
        if play[i] == "Yes":
            rain_yes += 1
        else:
            rain_no += 1

sunny_prob_yes = sunny_yes/sunny
sunny_prob_no = sunny_no/sunny
overcast_prob_yes = overcast_yes/overcast
overcast_prob_no = overcast_no/overcast
rain_prob_yes = rain_yes/rain
rain_prob_no = rain_no/rain

sunny_entropy = entropy(sunny_prob_yes, sunny_prob_no)
overcast_entropy = entropy(overcast_prob_yes, overcast_prob_no)
rain_entropy = entropy(rain_prob_yes, rain_prob_no)

outlook_entropy = (sunny/total_count)*sunny_entropy + (overcast/total_count)*overcast_entropy + (rain/total_count)*rain_entropy
outlook_IG = round((play_entropy - outlook_entropy), 3)
print("Outlook Entropy: ", round(outlook_entropy, 5))
print("Information Gain: ", outlook_IG )

Outlook Entropy:  0.69354
Information Gain:  0.247


In [153]:
data_temp = df3.filter(["Temperature", "Play?"])

hot = 0
hot_yes = 0
hot_no = 0
cool = 0
cool_yes = 0
cool_no = 0
mild = 0
mild_yes = 0
mild_no = 0

array = data_temp.values
temp = array[: , 0]
play = array[: , 1]

for i in range(0, total_count):
    if temp[i] == "Hot":
        hot += 1
        if play[i] == "Yes":
            hot_yes += 1
        else:
            hot_no += 1
    elif temp[i] == "Cool":
        cool += 1
        if play[i] == "Yes":
            cool_yes += 1
        else:
            cool_no += 1
    else:
        mild += 1
        if play[i] == "Yes":
            mild_yes += 1
        else:
            mild_no += 1
            
            
hot_prob_yes = hot_yes/hot
hot_prob_no = hot_no/hot
cool_prob_yes = cool_yes/cool
cool_prob_no = cool_no/cool
mild_prob_yes = mild_yes/mild
mild_prob_no = mild_no/mild

hot_entropy = entropy(hot_prob_yes, hot_prob_no)
cool_entropy = entropy(cool_prob_yes, cool_prob_no)
mild_entropy = entropy(mild_prob_yes, mild_prob_no)

temp_entropy = (hot/total_count)*hot_entropy + (cool/total_count)*cool_entropy + (mild/total_count)*mild_entropy
temp_IG = round((play_entropy - temp_entropy), 3)
print("Temperature Entropy: ", round(temp_entropy, 5))
print("Information Gain: ", temp_IG)

Temperature Entropy:  0.91106
Information Gain:  0.029


In [154]:
data_humid = df3.filter(["Humidity", "Play?"])

high = 0
high_yes = 0
high_no = 0
normal = 0
normal_yes = 0
normal_no = 0

array = data_humid.values
humid = array[: , 0]
play = array[: , 1]

for i in range(0, total_count):
    if humid[i] == "High":
        high += 1
        if play[i] == "Yes":
            high_yes += 1
        else:
            high_no += 1
    else:
        normal += 1
        if play[i] == "Yes":
            normal_yes += 1
        else:
            normal_no += 1
            
            
high_prob_yes = high_yes/high
high_prob_no = high_no/high
normal_prob_yes = normal_yes/normal
normal_prob_no = normal_no/normal

high_entropy = entropy(high_prob_yes, high_prob_no)
normal_entropy = entropy(normal_prob_yes, normal_prob_no)

humid_entropy = (high/total_count)*high_entropy + (normal/total_count)*normal_entropy
humid_IG = round((play_entropy - humid_entropy), 3)
print("Humidity Entropy: ", round(humid_entropy, 5))
print("Information Gain: ", humid_IG)

Humidity Entropy:  0.78845
Information Gain:  0.152


In [155]:
data_wind = df3.filter(["Windy", "Play?"])

false = 0
false_yes = 0
false_no = 0
true = 0
true_yes = 0
true_no = 0

array = data_wind.values
wind = array[: , 0]
play = array[: , 1]

for i in range(0, total_count):
    if wind[i] == "False":
        false += 1
        if play[i] == "Yes":
            false_yes += 1
        else:
            false_no += 1
    else:
        true += 1
        if play[i] == "Yes":
            true_yes += 1
        else:
            true_no += 1
            
false_prob_yes = false_yes/false
false_prob_no = false_no/false
true_prob_yes = true_yes/true
true_prob_no = true_no/true

false_entropy = entropy(false_prob_yes, false_prob_no)
true_entropy = entropy(true_prob_yes, true_prob_no)

wind_entropy = (false/total_count)*false_entropy + (true/total_count)*true_entropy 
wind_IG = round((play_entropy - wind_entropy), 3)
print("Windy Entropy: ",  round(wind_entropy, 5))
print("Information Gain: ", wind_IG)

Windy Entropy:  0.89216
Information Gain:  0.048


In [158]:
final_infoGain = {"Outlook" : outlook_IG, "Temperature" : temp_IG,
                 "Humidity" : humid_IG, "Windy" : wind_IG}

root = max(final_infoGain, key = final_infoGain.get)
print("Maximum Information Gain: ", root)

Maximum Information Gain:  Outlook
